File: Week 2 - Arrays/ProblemSet2/Readability/readability.c
/*
https://cs50.harvard.edu/x/2022/psets/2/readability/
*/

#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Get user input
    string text = get_string("Text: ");

    // Variables for letters, words, sentences
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Calculate average number of letters & average number of sentences
    float avgLetters = (100 / (float)words) * (float)letters;
    float avgSentences = (100 / (float)words) * (float)sentences;

    // Calculate Coleman-Liau index
    int index = round(0.0588 * avgLetters - 0.296 * avgSentences - 15.8);

    // Print out index based on result
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

// Determine the number of letters in a string
int count_letters(string text)
{
    int letters = 0;

    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }

    return letters;
}

// Determine the number of words in a string
int count_words(string text)
{
    int words = 1;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == ' ')
        {
            words++;
        }
    }

    return words;
}

// Determine the number of sentences in a string
int count_sentences(string text)
{
    int sentences = 0;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }

    return sentences;
}

File: Week 6 - Python/ProblemSet6/SentimentalReadability/readability.py
"""
https://cs50.harvard.edu/x/2023/psets/6/readability/
"""

def main():

    sentence = input("Text: ")

    # Number of letters, words and sentences in the string.
    letters = count_letters(sentence)
    words = count_words(sentence)
    sentences = count_sentences(sentence)

    # Average number of letters and stentences per 100 words in text.
    avg_letters = (100 / float(words)) * float(letters)
    avg_sentences = (100 / float(words)) * float(sentences)

    # Coleman-Liau index.
    index = calculate_index(avg_letters, avg_sentences)

    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


def count_letters(sentence):
    """Counts and returns the number of letters in a string."""

    letters = 0

    for character in sentence:
        if character.isalpha():
            letters += 1

    return letters


def count_words(sentence):
    """Counts and returns the number of words in a string."""

    return len(sentence.split())


def count_sentences(sentence):
    """Counts and returns the number of sentences in a string."""

    sentences = 0

    for character in sentence:
        if character in [".", "!", "?"]:
            sentences += 1

    return sentences


def calculate_index(avg_letters, avg_sentences):
    """Calculates the approximate grade level needed to comprehend some text according to the Coleman-Liau formula. Returns the result of the formula."""

    return round(0.0588 * avg_letters - 0.296 * avg_sentences - 15.8)


if __name__ == "__main__":
    main()
